### PR TITLE
Add support for Square Register

### PR DESCRIPTION
**Pull request by cthorne91**
- Added square session manager
- Updated the demo project.

I didn't have client ids and secrets to give out like the other services have, so as users try out the demo they will have to enter their own credentials. Hope that's ok.

### CODE REVIEW

File: Sources/WFOAuth2/WFSquareOAuth2SessionManager.m [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    NSArray<NSURLQueryItem *> *parameters = @[[NSURLQueryItem queryItemWithName:@"client_id" value:self.clientID],
+                                              [NSURLQueryItem queryItemWithName:@"access_token" value:credential.accessToken]];
+    
+    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:[NSURL URLWithString:@"https://connect.squareup.com/oauth2/revoke"]];
+    [request setHTTPMethod:@"POST"];
+    
+    [request wfo_setBodyWithQueryItems:parameters];
+    
+    if (self.clientSecret)
+        [request addValue:[NSString stringWithFormat:@"Authorization: Client %@", self.clientSecret] forHTTPHeaderField:@"Authorization"];
```
Comment (conradev) [inline Sources/WFOAuth2/WFSquareOAuth2SessionManager.m]: I assume you meant `-[NSMutableURLRequest setValue: forHTTPHeaderField:]`?

File: Sources/WFOAuth2/WFOAuth2SessionManager.m [hunk 2]
```diff
@@ -256,8 +256,15 @@ - (WKWebView *)authorizationWebViewWithURL:(NSURL *)authorizationURL
         if ([redirectComponents.path hasSuffix:@"/"])
             redirectComponents.path = [redirectComponents.path substringToIndex:(redirectComponents.path.length - 1)];
         
+        /*
+            Square requires you to save the redirect URL in the webportal, there is no sense in passing it in as a parameter.
+            The assumption is made that redirecting to something other than the square domain is considered redirecting to your redirect url.
+         */
+        if (!redirectURI && [components.host containsString:@"squareup"])
```
Comment (conradev) [inline Sources/WFOAuth2/WFOAuth2SessionManager.m]: > Square requires you to save the redirect URL in the webportal, there is no sense in passing it in as a parameter.

We pass it as a parameter for two reasons:
1. Many services (not Square, it seems) use it to verify if a request is valid
2. It helps us identify which navigation actions signify completion

Unless there is a way to deal with `2` in a clean way I am hesitant to merge this as is. I'd consider a clean way to be something like this:
- Containing this peculiarity to the Square subclass somehow
- Recognizing completion for a `nil` `redirectURI` in a more generic way (determining if the host of the website changes, or scanning all URLs for the completion parameters)

**Comment by conradev** [comment]
Thank you for taking the time to put this together! Besides the two comments I made, this looks really great. Let me know what you think and we can get this merged.

Also, I can go ahead and do it after it is merged, but having this added to the `README` would be 💯

**Comment by conradev** [comment]
Re: not having a client ID and secret for the demo project, I just created #6 to solve that

File: Sources/WFOAuth2/WFOAuth2SessionManager.m (hunk 2 above)
Comment (cthorne91) [inline Sources/WFOAuth2/WFOAuth2SessionManager.m]: Yes, I was hesitant with this section as well. I should have cleaned it up from the start, sorry. I like your idea of scanning for completion parameters. As I understand the OAuth 2.0 spec, the required parameters in our case would be `code` and `state`. In the case of an error the sole required parameter would be `error`. I'll scan for those in the case of a `nil` `redirectURI`. I'll update the branch shortly.

File: Sources/WFOAuth2/WFSquareOAuth2SessionManager.m (hunk 1 above)
Comment (cthorne91) [inline Sources/WFOAuth2/WFSquareOAuth2SessionManager.m]: Code review = good
Thanks. I'll update that.

**Comment by cthorne91** [comment]
No problem, thanks for the repo!

I would love to update the `README` but I don't have a place to host the icon if that's what you mean. I can however save you the trouble of finding this.
square_icon

**Comment by conradev** [comment]
Awesome, was able to test this out and everything looks good code-wise! If you could make the following stylistic changes, I can get this merged ASAP:
1. Change the `@"Ok"` in `-[ViewController squarePressed:]` to `@"OK"` (that's the iOS standard dialog message)
2. Change the parentheses in the ternary operator in the decision handler to match `(redirectURI ? … : … )`
3. Remove the comment about `nil` `redirectURI` in the decision handler (see below)
4. Change the code block in the handler at the end to be the following, for clarity:

``` objc
If (redirectURI) {
    // Uber redirects
    …
} else {
    …
}
```
1. Squash all commits into the final three commits:
   - `Add Square session manager subclass`
   - `Remove assert on nullable parameter`
   - `Support nil redirectURI in authorization web views` (with a short description explaining the change in the commit message)

Let me know if you need any clarification. Looking forward to getting this merged 🎉

**Comment by cthorne91** [comment]
I've pushed those changes. Let me know if everything is how you expect it to be. If you need anything else before the merge let me know. 👍🏼

**Comment by conradev** [comment]
It appears as if the commits were squashed incorrectly into one other:
    - The `Remove assert on nullable parameter` commit should only have that one line change
    - The `Support nil redirectURI in authorization web views` commit should have all of the changes to `WFOAuth2SessionManager` and nothing else
    - Everything relating to Square should be in `Add Square session manager subclass` commit.

Also, from above, the

``` objc
if (redirectURI) {
    …
} else {
    …
}
```

block should include everything at the end of the handler, and there should be no code outside/after it.

**Comment by cthorne91** [comment]
Thanks for your patience with me. This is my first time really doing a rebase and of course I messed it up. I'll fix that and take care of the `if/else` issue as well.

**Comment by conradev** [comment]
No worries at all! I've totally been there haha. I can do it for you if you'd like help – there's probably some `git` gymnastics required :)

**Comment by cthorne91** [comment]
I think I got it worked out. Let me know if the commit history is accurate now. If not, sure let me know what I need to do to let you help me out.
Thanks

**Comment by conradev** [comment]
Did you forget to push it? The commits in the pull request are still showing up as the old ones.

**Comment by cthorne91** [comment]
I did push. I probably just didn't do the rebase correctly or something. Do you see that commit hash `08adbef` titled: 'Remove assert on nullable parameter' has only one change? It used to have a lot more. I'm seeing that it only has 1 change now as requested, for example. Are you referring to something else?

I don't want to waste too much of your time, so if there is anything I can do to save your time let me know!

**Comment by conradev** [comment]
Awesome, yeah, `08adbef` looks great.

`Support nil redirectURI in authorization web views` still has the Square demo mode and the `setValue:forHTTPHeader:` fix and other stuff. It should only have changes to `WFOAuth2SessionManager`.

**Comment by cthorne91** [comment]
Ok, you were right. That was some `git` gymnastics. I believe the commits are structured appropriately now. Let me know.

**Comment by conradev** [comment]
Wooo, looks great! Last thing, and I can click the green button: the explanation that is supposed to be in `8bfcbe4` is missing

**Comment by cthorne91** [comment]
Ok, updated. Thanks for your help and patience! 
Much appreciated.

**Comment by conradev** [comment]
Of course! Thanks again for contributing. Congrats on your first rebase! 😄 

&nbsp;&nbsp;&nbsp;&nbsp;☁️ 
&nbsp;☁️ 
🚂🚋🚋🚋🚋🚋🚋🚋

**Comment by Bogidon** [comment]
@conradev I think you forgot to make a new release for this PR. Shipping it out soon?

**Pull request merged**